components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+ */
+@property(nonatomic, assign) CGFloat preferredMainContentAreaHeight;
+
+/**
+ This API allows the user to override the default underline label area height. The underline label
+ area is the part of the view where the underline labels are. It is located below the main content
+ area. If this property is set to a value that's lower than the default underline label area height
+ the value will be ignored in the calculation of the view's @c intrinsicContentSize.
+ */
+@property(nonatomic, assign) CGFloat preferredUnderlineLabelAreaHeight;
+
+@end
+
+/**
+ This protocol represents a set of colors that are semantically meaningful and specific to
+ MDCContainedInputView. Each property corresponds to the color of one or more views that an
+ MDCContainedInputView manages at a given point of time.
+ */
+@protocol MDCContainedInputViewColorScheming <NSObject>
+/**
+ The color of the contained input view's text.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *textColor;
+/**
+ The color of the contained input view's underline label.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *underlineLabelColor;
+/**
+ The color of the contained input view's floating label.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *floatingLabelColor;
+/**
+ The color of the contained input view's placeholder label.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *placeholderColor;
+/**
+ The tint color of the contained input view's clear button.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *clearButtonTintColor;
+/**
+ The color the contained input view should apply during an error state.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *errorColor;
+@end
+
+/**
+ A base implementation of MDCContainedInputViewColorScheming. Intended to be subclassed for styles
+ that include additional elements that need to be colored.
+ */
+@interface MDCContainedInputViewColorScheme : NSObject <MDCContainedInputViewColorScheming>
+@property(strong, nonatomic, nonnull) UIColor *textColor;
+@property(strong, nonatomic, nonnull) UIColor *underlineLabelColor;
+@property(strong, nonatomic, nonnull) UIColor *floatingLabelColor;
+@property(strong, nonatomic, nonnull) UIColor *placeholderColor;
+@property(strong, nonatomic, nonnull) UIColor *clearButtonTintColor;
+@property(strong, nonatomic, nonnull) UIColor *errorColor;
+@end
+
+@protocol MDCContainedInputViewStyleDensityInforming;
+
+@protocol MDCContainedInputViewStyle <NSObject>
+/**
+ The style's densityInformer. Different implementations of MDCContainedInputView may need different
+ density informers on a per style basis.
+ */
+@property(strong, nonatomic, nonnull) id<MDCContainedInputViewStyleDensityInforming>
+    densityInformer;
+/**
+ This method provides a default object conforming to MDCContainedInputViewColorScheming.
+ */
+- (nonnull id<MDCContainedInputViewColorScheming>)defaultColorSchemeForState:
+    (MDCContainedInputViewState)state;
```

Comment:
The convention in MDC components is that there's a default theming that's hard coded in the component. Than there's MDC baseline themers and GM themers (ie: theming extension).  Which of these is this one? Why does it need its own API?

Reply:
The styling itself is not built into any object conforming to MDCContainedInputView, so a contained input view doesn't necessarily know what needs to be colored. The style knows. This API allows the style to tell the contained input view. The result of this is that if someone did something like `containedInputView.containerStyle = aNewContainerStyle` the view would know what needs to be colored and how to color it.